File: python/mlx_sfx_model_paths.py
"""解析 MOSS-SoundEffect 本地 MLX 权重目录。"""
from __future__ import annotations

import json
from pathlib import Path

_SFX_SUBDIR_NAMES = (
    "mlx-4bit",
    "mlx-int8",
    "mlx",
    "MOSS-SoundEffect-MLX-4bit",
    "MOSS-SoundEffect",
    "openmoss-sound-effect-mlx",
)


def dir_has_mlx_sfx_weights(path: Path) -> bool:
    if not path.is_dir():
        return False
    if not (path / "config.json").is_file():
        return False
    if (path / "model.safetensors").is_file():
        return True
    if (path / "model.safetensors.index.json").is_file():
        return True
    return any(p.suffix == ".safetensors" for p in path.iterdir() if p.is_file())


def dir_looks_hf_only(path: Path) -> bool:
    if not path.is_dir() or not (path / "config.json").is_file():
        return False
    if dir_has_mlx_sfx_weights(path):
        return False
    if (path / "pytorch_model.bin").is_file():
        return True
    try:
        cfg = json.loads((path / "config.json").read_text(encoding="utf-8"))
        mt = str(cfg.get("model_type", "")).lower()
        if mt and "moss" not in mt and "delay" not in mt:
            return True
    except (OSError, json.JSONDecodeError):
        pass
    return False


def resolve_mlx_sfx_model_dir(model_root: str | Path) -> Path | None:
    root = Path(model_root).expanduser().resolve()
    if not root.is_dir():
        return None
    if dir_has_mlx_sfx_weights(root):
        return root
    for name in _SFX_SUBDIR_NAMES:
        cand = root / name
        if dir_has_mlx_sfx_weights(cand):
            return cand
    try:
        for child in sorted(root.iterdir()):
            if child.is_dir() and dir_has_mlx_sfx_weights(child):
                return child
    except OSError:
        pass
    return None


def diagnose_sfx_model_root(model_root: str | Path) -> dict:
    root = Path(model_root).expanduser().resolve()
    resolved = resolve_mlx_sfx_model_dir(root)
    if resolved is not None:
        return {"ok": True, "resolved": str(resolved), "root": str(root)}
    if dir_looks_hf_only(root):
        return {
            "ok": False,
            "kind": "hf_only",
            "root": str(root),
            "message": (
                "检测到非 MLX 权重（如 pytorch_model.bin）。"
                "请下载 mlx-community/MOSS-SoundEffect-MLX-4bit 并指向含 safetensors 的目录。"
            ),
        }
    return {
        "ok": False,
        "kind": "not_found",
        "root": str(root),
        "message": (
            "未找到含 config.json 与 *.safetensors 的 MLX 权重。"
            "请指向 ModelScope：mlx-community/MOSS-SoundEffect-MLX-4bit（或其中 mlx-4bit 子目录）。"
        ),
    }

File: electron/ai-model-service/adapters/mlxSfxModelPaths.ts
/**
 * 解析 MOSS-SoundEffect 本地权重目录（与 python/mlx_sfx_model_paths.py 对齐）。
 */
import fs from 'node:fs';
import path from 'node:path';

const SFX_SUBDIR_NAMES = [
  'mlx-4bit',
  'mlx-int8',
  'mlx',
  'MOSS-SoundEffect-MLX-4bit',
  'MOSS-SoundEffect',
  'openmoss-sound-effect-mlx',
] as const;

export function dirHasMlxSfxWeights(dir: string): boolean {
  if (!fs.existsSync(dir) || !fs.statSync(dir).isDirectory()) return false;
  if (!fs.existsSync(path.join(dir, 'config.json'))) return false;
  if (fs.existsSync(path.join(dir, 'model.safetensors'))) return true;
  if (fs.existsSync(path.join(dir, 'model.safetensors.index.json'))) return true;
  try {
    return fs.readdirSync(dir).some((f) => f.endsWith('.safetensors'));
  } catch {
    return false;
  }
}

export function dirLooksHfOnlySfx(dir: string): boolean {
  if (!fs.existsSync(dir) || !fs.statSync(dir).isDirectory()) return false;
  if (!fs.existsSync(path.join(dir, 'config.json'))) return false;
  if (dirHasMlxSfxWeights(dir)) return false;
  return fs.existsSync(path.join(dir, 'pytorch_model.bin'));
}

export function resolveMlxSfxModelDir(modelRoot: string): string | null {
  const root = path.resolve(modelRoot.trim());
  if (!fs.existsSync(root) || !fs.statSync(root).isDirectory()) return null;
  if (dirHasMlxSfxWeights(root)) return root;
  for (const name of SFX_SUBDIR_NAMES) {
    const cand = path.join(root, name);
    if (dirHasMlxSfxWeights(cand)) return cand;
  }
  try {
    for (const name of fs.readdirSync(root).sort()) {
      const cand = path.join(root, name);
      if (fs.statSync(cand).isDirectory() && dirHasMlxSfxWeights(cand)) return cand;
    }
  } catch {
    /* ignore */
  }
  return null;
}

export function diagnoseSfxModelRoot(modelRoot: string): {
  ok: boolean;
  resolved?: string;
  root: string;
  message?: string;
} {
  const root = path.resolve((modelRoot ?? '').trim());
  const resolved = resolveMlxSfxModelDir(root);
  if (resolved) {
    return { ok: true, resolved, root };
  }
  if (dirLooksHfOnlySfx(root)) {
    return {
      ok: false,
      root,
      message:
        '检测到非 MLX 权重。请下载 mlx-community/MOSS-SoundEffect-MLX-4bit 并指向含 safetensors 的目录。',
    };
  }
  return {
    ok: false,
    root,
    message:
      '未找到 MLX 权重（config.json + safetensors）。请指向 MOSS-SoundEffect-MLX-4bit 或其中 mlx-4bit 子目录。',
  };
}
